### ISSUE TITLE
4.0.2 > 4.1.0 breaking changes

### ISSUE THREAD

**Issue by RWOverdijk** [marmelab/react-admin#7725]
**What you were expecting:**

I just updated to 4.1.0 and notice some breaking changes.

1. The permission function provided as a child to <Admin /> now gets called with undefined the first time (broke my dashboard).
2. It also requires the function to be a promise now.

Both of these weren't an issue on 4.0.2.

I fixed these issues easily by adding a simple `if (!p) return null` at the top of the function and making it async. Visually:

```tsx
<Admin>
  {async permissions => { // < added async
    if (!permissions) return null; // < added check

    // ...
  }}
</Admin>
```

The reason I am still writing this issue is because this version is tagged as a minor version and so it should have been backwards compatible. I thought you might want to know.

**Update:** I just tested the rest and all my other issues were fixed in this release though, so I am actually really happy with 4.1.0

**Comment by fzaninotto** [comment]
Hi, and thanks for your feedback. This is likely linked to changes introduced in 4.0.3 (#7606, #7609).

Regarding your second point, I can't reproduce it in the simple example codesandbox, which uses a function that does not return a Promise. 

https://codesandbox.io/s/github/marmelab/react-admin/tree/master/examples/simple

Could you please build a repro based on that CodeSandbox? 

Regarding your first point, I'm not sure this is a bug - we'll also need a repro to see.

Thanks in advance

**Comment by RWOverdijk** [comment]
@fzaninotto I can't (unfortunately no time). But:

> useConfigureAdminRouterFromChildren.tsx:116 TypeError: Cannot read properties of null (reading 'then')

![image](https://user-images.githubusercontent.com/781745/169522009-0d70b8a1-cee5-405b-a074-82cf2cae2557.png)

this is in useConfigureAdminRouterFromChildren.tsx

I don't know why the sandbox doesn't have the issue.

Feel free to close this if this isn't enough info, no hard feelings 😄

**Comment by fzaninotto** [comment]
That's the problem: the function shouldn't return null, but a list of children. I believe this is due to the fact that, if you don't have permissions, you return nothing. 

I managed to reproduce the second error : https://codesandbox.io/s/wonderful-lichterman-lez2zo?file=/src/authProvider.tsx

So I'm marking this as a bug - But I'm not sure what the admin should show when the function returns nothing.

**Comment by RWOverdijk** [comment]
@fzaninotto The first call to the permissions function (with undefined) used to not happen, I think that's why I noticed the null issue now. If you don't have permissions you're not allowed to see any of the resources. When you log in, I can finally start filtering resources. But returning null from a promise is allowed and doesn't break, so it's not fatal. Just interesting 😄

**Comment by RWOverdijk** [comment]
Hm, something does seem broken. It's no longer redirecting me to /login when I'm not logged in, for some reason. I'll reply here once I find out what the problem is.

**Comment by fzaninotto** [comment]
I'm reopening the issue as @djhi 's fix only concerns regression #2 - regression #1 is still not fixed.
